moveit_core/collision_detection_fcl/src/collision_common.cpp

```diff
@@ -592,53 +592,90 @@ bool distanceDetailedCallback(fcl::CollisionObject* o1, fcl::CollisionObject* o2
   {
-    it1 = cdata->res->distance.find(cd1->ptr.obj->id_);
-    it2 = cdata->res->distance.find(cd2->ptr.obj->id_);
+    it1 = cdata->res->distances.find(cd1->ptr.obj->id_);
+    it2 = cdata->res->distances.find(cd2->ptr.obj->id_);
 
     if (cdata->req->active_components_only)
     {
       if (active1 && active2)
       {
-        if (it1 != cdata->res->distance.end() && it2 != cdata->res->distance.end())
-          dist_threshold = std::max(it1->second.min_distance, it2->second.min_distance);
+        if (it1 != cdata->res->distances.end() && it2 != cdata->res->distances.end())
+          dist_threshold = std::max(it1->second.distance, it2->second.distance);
       }
       else if (active1 && !active2)
       {
-        if (it1 != cdata->res->distance.end())
-          dist_threshold = it1->second.min_distance;
+        if (it1 != cdata->res->distances.end())
+          dist_threshold = it1->second.distance;
       }
       else if (!active1 && active2)
       {
-        if (it2 != cdata->res->distance.end())
-          dist_threshold = it2->second.min_distance;
+        if (it2 != cdata->res->distances.end())
+          dist_threshold = it2->second.distance;
       }
     }
     else
     {
-      if (it1 != cdata->res->distance.end() && it2 != cdata->res->distance.end())
-        dist_threshold = std::max(it1->second.min_distance, it2->second.min_distance);
+      if (it1 != cdata->res->distances.end() && it2 != cdata->res->distances.end())
+        dist_threshold = std::max(it1->second.distance, it2->second.distance);
     }
   }
   else
   {
-    dist_threshold = cdata->res->minimum_distance.min_distance;
+    dist_threshold = cdata->res->minimum_distance.distance;
   }
 
   fcl_result.min_distance = dist_threshold;
-  double d = fcl::distance(o1, o2, fcl::DistanceRequest(cdata->req->detailed), fcl_result);
+  double d = fcl::distance(o1, o2, fcl::DistanceRequest(cdata->req->enable_nearest_points), fcl_result);
 
   // Check if either object is already in the map. If not add it or if present
   // check to see if the new distance is closer. If closer remove the existing
   // one and add the new distance information.
   if (d < dist_threshold)
   {
-    dist_result.min_distance = fcl_result.min_distance;
+    dist_result.distance = fcl_result.min_distance;
     dist_result.nearest_points[0] = Eigen::Vector3d(fcl_result.nearest_points[0].data.vs);
     dist_result.nearest_points[1] = Eigen::Vector3d(fcl_result.nearest_points[1].data.vs);
-    dist_result.link_name[0] = cd1->ptr.obj->id_;
-    dist_result.link_name[1] = cd2->ptr.obj->id_;
+    dist_result.link_names[0] = cd1->ptr.obj->id_;
+    dist_result.link_names[1] = cd2->ptr.obj->id_;
+    dist_result.hasNearestPoints = cdata->req->enable_nearest_points;
+    if (dist_result.hasNearestPoints)
+    {
+      dist_result.normal = (dist_result.nearest_points[1] - dist_result.nearest_points[0]).normalized();
+    }
+
+    if (d <= 0 && cdata->req->enable_signed_distance)
```

Comment:
Why do you need this extra code block? Shouldn't `fcl::distance` compute the correct distance?

Reply:
That is what I had hoped, but in fcl 0.5 if two objects are in collision it returns 0 or -1 depending on the object type. Now fcl has been making a lot of improvements over the past year and in version 0.6 which has not been release will return this information directly. This interim fix to get this information until the 0.6 release. I also have another branch which I will push this week that make the necessary change for 0.6 since they reorganized everything and switch to using Eigen instead of custom matrix types.

Reply:
I see.